iceoryx_utils/include/iceoryx_utils/cxx/unique_ptr.hpp

```diff
@@ -0,0 +1,125 @@
+
+namespace iox
+{
+namespace cxx
+{
+///
+/// @todo document how it differs to STL
+///
+template <typename T>
+class unique_ptr
+{
+  public:
+    using ptr_t = T*;
+
+    unique_ptr() = delete;
+
+    ///
+    /// @brief unique_ptr Creates an empty unique ptr that owns nothing. Can be passed ownership later via reset.
+    ///
+    unique_ptr(std::function<void(T*)>&& deleter) noexcept;
+
+    ///
+    /// @brief unique_ptr Creates a unique pointer that takes ownership of an object.
+    /// @details A deleter must always be provided as no default can be provided given that no head is used.
+    /// The unique_ptr must know how to delete the managed object when pointer out of scope.
+    /// @param ptr The raw pointer to the object to be managed.
+    /// @param deleter The deleter function for cleaning up the managed object.
+    ///
+    unique_ptr(ptr_t ptr, std::function<void(T*)>&& deleter) noexcept;
+
+    unique_ptr(std::nullptr_t) noexcept;
+
+    // Not copy-able to ensure uniqueness.
+    unique_ptr(const unique_ptr& other) = delete;
+    unique_ptr& operator=(const unique_ptr&) = delete;
+
+    unique_ptr(unique_ptr&& rhs) noexcept;
+    unique_ptr& operator=(unique_ptr&& rhs) noexcept;
+
+    ///
+    /// @brief operator = Reset to empty pointer when setting to nullptr.
+    /// @return An empty unique pointer.
+    ///
+    unique_ptr& operator=(std::nullptr_t) noexcept;
+
+    ///
+    /// Automatically deletes the owned object on destruction.
+    ///
+    ~unique_ptr() noexcept;
+
+    ///
+    /// Dereference the stored pointer.
+    ///
+    T operator*() noexcept;
+
+    ///
+    /// Return the stored pointer.
+    ///
+    ptr_t operator->() noexcept;
+
+    ///
+    /// @brief operator bool Returns true if it points to something.
+    ///
+    explicit operator bool() const noexcept
+    {
+        return get() == ptr_t() ? false : true;
+    }
+
+    ///
+    /// @brief get Retrieve the underlying raw pointer.
+    /// @details The unique_ptr retains ownership, therefore the "borrowed" pointer must not be deleted.
+    /// @return Pointer to managed object or nullptr if none owned.
+    ///
+    ptr_t get() const noexcept;
+
+    ///
+    /// @brief release Release ownership of the underlying pointer.
+    /// @return Pointer to the managed object or nullptr if none owned.
+    ///
+    ptr_t release() noexcept;
+
+    ///
+    /// @brief reset Reset the unique pointer to take ownership of the given pointer.
+    /// @details Any previously owned objects will be deleted. If no pointer given then points to nullptr.
+    /// @param ptr Pointer to object to take ownership on.
+    ///
+    void reset(ptr_t ptr = nullptr) noexcept;
+
+    ///
+    /// @brief swap Swaps object ownership with another unique_ptr.
+    /// @param other The unique_ptr with which to swap owned objects.
+    ///
+    void swap(unique_ptr& other) noexcept;
+
+  private:
+    ptr_t m_ptr = nullptr;
+    std::function<void(T* const)> m_deleter;
```

Comment:
In our use case we can circumvent the deleter problem. Let's say we have a free template function like:
```cpp
template<typename T>
void FreeStuff(T * const ptr, const Publisher & publisher) {
    publisher.freeSample(ptr);
}
```

Then we could create a unique ptr like
```cpp
cxx::unique_ptr<int> bla(ptrToInt, FreeStuff);
```
and still use a `cxx::function_ref` since this free function exists as long as the program is running. But for this to work the unique ptr needs to store the pointer of its origin. We already realized a similar approach in our shared chunk  and shared_ptr. But it would mean that the unique ptr is not generic anymore and has to move into popo.

Reply:
This idea should be explored when doing #298.